From 657e5e9aa6600cbb40e42432e2e3b6b50678de63 Mon Sep 17 00:00:00 2001
From: rzou <zou3519@gmail.com>
Date: Mon, 28 Jul 2025 13:08:00 -0700
Subject: [PATCH] All custom operators go through Inductor's
 graph.call_function (#159174)

Fixes #158892

All custom operators should go through the graph.call_function path. The
other fallback path is for aten/prim operations that don't have support
for things (like torch.float8_e8m0fn).

Test Plan:
- new tests

Pull Request resolved: https://github.com/pytorch/pytorch/pull/159174
Approved by: https://github.com/eellison
---
 test/inductor/test_needs_exact_strides.py | 102 ++++++++++++++++++++++
 torch/_inductor/graph.py                  |   4 +-
 2 files changed, 105 insertions(+), 1 deletion(-)
 create mode 100644 test/inductor/test_needs_exact_strides.py

diff --git a/test/inductor/test_needs_exact_strides.py b/test/inductor/test_needs_exact_strides.py
new file mode 100644
index 0000000000000..97c760c712bf2
--- /dev/null
+++ b/test/inductor/test_needs_exact_strides.py
@@ -0,0 +1,102 @@
+# Owner(s): ["module: inductor"]
+
+import torch
+import torch.utils._pytree as pytree
+from torch._inductor.pattern_matcher import (
+    CallFunctionVarArgs,
+    PatternMatcherPass,
+    register_graph_pattern,
+)
+from torch._inductor.test_case import run_tests, TestCase as InductorTestCase
+from torch.testing._internal.common_utils import (
+    instantiate_parametrized_tests,
+    IS_LINUX,
+    parametrize,
+)
+from torch.testing._internal.inductor_utils import HAS_GPU
+
+
+class TestNeedsExactStrides(InductorTestCase):
+    @parametrize("dtype", [torch.float, torch.float8_e8m0fnu])
+    def test_custom_op(self, dtype):
+        device = "cuda"  # float8_e8m0fnu errors on "cpu"
+        x = torch.ones(4, 4, 2, 2, device=device, dtype=torch.float8_e8m0fnu)
+        other = torch.ones(4, 4, 2, 2, device=device, dtype=torch.float8_e8m0fnu)
+
+        class _CustomPass(PatternMatcherPass):
+            def __init__(self) -> None:
+                super().__init__()
+
+            def __call__(self, g: torch.fx.Graph):
+                self.apply(g)
+
+        g = _CustomPass()
+        called = False
+
+        @register_graph_pattern(
+            CallFunctionVarArgs(torch.ops.aten.permute),
+            pass_dict=g,
+        )
+        def _(match, *args, **kwargs):
+            flat_args, spec = pytree.tree_flatten((args, kwargs))
+
+            def decomp(*flat_args):
+                args, kwargs = pytree.tree_unflatten(flat_args, spec)
+                return torch.ops.mylib.force_channels_last(
+                    torch.ops.aten.permute(*args, **kwargs)
+                )
+
+            nonlocal called
+            called = True
+            match.replace_by_example(decomp, flat_args)
+
+        from torch._inductor import config
+
+        class TestPassed(RuntimeError):
+            pass
+
+        with torch.library._scoped_library("mylib", "FRAGMENT") as lib:
+            lib.define(
+                "force_channels_last(Tensor x) -> Tensor",
+                tags=[torch._C.Tag.flexible_layout],
+            )
+
+            def impl2(x):
+                return x.clone(memory_format=torch.channels_last)
+
+            lib.impl("force_channels_last", impl2, "CompositeExplicitAutograd")
+
+            lib.define(
+                "add_op(Tensor x, Tensor y) -> Tensor",
+            )
+
+            def impl(x, y):
+                assert x.transpose(2, 3).is_contiguous()
+                assert y.is_contiguous()
+                return x.float() + y.float()
+
+            def meta(x, y):
+                return x.float() + y.float()
+
+            lib.impl("add_op", impl, "CompositeExplicitAutograd")
+            lib.impl("add_op", meta, "Meta")
+
+            def f(x, other):
+                return torch.ops.mylib.add_op.default(x.transpose(2, 3), other)
+
+            with config.patch(
+                post_grad_custom_post_pass=g,
+            ):
+                try:
+                    f_compile = torch.compile(f, fullgraph=True)
+                    f_compile(x, other)
+                except TestPassed:
+                    pass
+                assert called
+
+
+instantiate_parametrized_tests(TestNeedsExactStrides)
+
+if __name__ == "__main__":
+    if IS_LINUX and HAS_GPU:
+        run_tests(needs="filelock")
diff --git a/torch/_inductor/graph.py b/torch/_inductor/graph.py
index 660b01b69233b..2f0670c75725c 100644
--- a/torch/_inductor/graph.py
+++ b/torch/_inductor/graph.py
@@ -1560,8 +1560,10 @@ def debug(msg: str) -> None:
         ):
             if (
                 n.op == "call_function"
+                # this path only for built-in operators
                 and n.target
-                not in (operator.getitem, torch._higher_order_ops.invoke_subgraph)
+                and isinstance(n.target, torch._ops.OpOverload)
+                and torch._library.utils.is_builtin(n.target)
                 and (
                     fallback_node_due_to_unsupported_type(n)
                     or CompilerBisector.disable_subsystem(
